internal/backends/sqlite/metadata/registry.go

```diff
@@ -353,6 +374,94 @@ func (r *Registry) CollectionRename(ctx context.Context, dbName, oldCollectionNa
 	panic("not implemented")
 }
 
+// indexesCreate creates an index in the collection.
+//
+// It does not hold the lock.
+func (r *Registry) indexesCreate(ctx context.Context, db *fsql.DB, c *Collection, indexes []IndexInfo) error {
+	err := db.InTransaction(ctx, func(tx *fsql.Tx) error {
+		for _, index := range indexes {
+			// check if index already exists
+			q := `SELECT EXISTS (SELECT 1 FROM sqlite_master WHERE type= 'index' and tbl_name = ? and name = ?)`
+			row := tx.QueryRowContext(ctx, q, c.TableName, c.TableName+index.Name)
+
+			var exists bool
+			err := row.Scan(&exists)
+			if err != nil {
+				return lazyerrors.Error(err)
+			}
+
+			if exists {
+				continue
+			}
+
+			q = "CREATE "
+
+			if index.Unique {
+				q += "UNIQUE "
+			}
+
+			q += "INDEX %q ON %q (%s)"
+
+			columns := make([]string, len(index.Key))
+			for i, key := range index.Key {
+				columns[i] = fmt.Sprintf("%s->'$.%s'", DefaultColumn, key.Field)
+				if key.Descending {
+					columns[i] += " DESC"
+				}
+			}
+
+			q = fmt.Sprintf(q, c.TableName+index.Name, c.TableName, strings.Join(columns, ", "))
+
+			if _, err := db.ExecContext(ctx, q); err != nil {
+				return lazyerrors.Error(err)
+			}
+
+			c.Settings.Indexes = append(c.Settings.Indexes, index)
+		}
+
+		q := fmt.Sprintf("UPDATE %q SET settings=?", metadataTableName)
+		if _, err := db.ExecContext(ctx, q, c.Settings); err != nil {
+			return lazyerrors.Error(err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return lazyerrors.Error(err)
+	}
+
+	return nil
+}
+
+// IndexesCreate creates an index in the collection.
+func (r *Registry) IndexesCreate(ctx context.Context, dbName, collectionName string, indexes []IndexInfo) error {
+	defer observability.FuncCall(ctx)()
+
+	r.rw.Lock()
+	defer r.rw.Unlock()
+
+	_, err := r.collectionCreate(ctx, dbName, collectionName)
+	if err != nil {
+		return lazyerrors.Error(err)
+	}
+
+	db, err := r.databaseGetOrCreate(ctx, dbName)
+	if err != nil {
+		return lazyerrors.Error(err)
+	}
+
+	c := r.collectionGet(dbName, collectionName)
+
+	err = r.indexesCreate(ctx, db, c, indexes)
+	if err != nil {
+		return lazyerrors.Error(err)
+	}
+
+	r.colls[dbName][collectionName] = c
```

Comment:
Should we call `r.collectionGet(dbName, collectionName)` after `r.indexesCreate(ctx, db, c, indexes)`?
I assume this line updates `Settings` of Collection.

Reply:
Yes, this row is to update settings. We can call `r.collectionGet`, but I'm not sure why. We already blocked everything with the mutex, so we don't need any extra check  that collection is updated.

Also please not that calling `indexCreate` modifies `c` (Settings are set). Maybe I need to document it.

Reply:
That line should be inside indexesCreate

Reply:
@AlekSi it can't be there. If `indexesCreate` was called from `collectionCreate` to create the default index, technically, collection doesn't exist before we exit from `collectionCreate`, so we can't update it from `indexesCreate`.

Reply:
> technically, collection doesn't exist before we exit from collectionCreate

I'm not sure why is that, but I'm sure that for any given function `foo()` I could write:

```go
func foo() {
  // ...
}

// For call foo() with lock.
func Foo() {
  l.Lock()
  defer l.Unlock()

  foo()
}
```

That's what we do for other exported/unexported function pairs, so I don't see the reason to not do the same there

Reply:
It's not enough. Collection creation consists of a few steps. One of the steps is index creation, another step is metadata insertion. Only after metadata insertion is complete correctly, we can consider collection is created and can be stored in `r.colls`. Otherwise, "everyone" is an owner and can modify `r.colls`, and this is not a safe approach.

Reply:
ok, I can copy-paste the same logic to `indexesCreate`, but I still don't think it's a good idea. two different things shouldn't both be owners. collection is already a good and suitable owner.

Reply:
The owner is the registry. `r.colls` is modified by many of its methods already, including `databaseDrop` that is not about collections.

And to be clear – I'm not asking you to copy&paste. All I'm saying that exported/unexported method pairs should be consistent.

Reply:
@AlekSi then I don't understand what you mean. Could you please suggest the exact changes? (either a commit or just a code suggestion).

Reply:
Just do it as close as possible for now